doc/manual/src/release-notes/rl-2.4.md

```diff
@@ -339,6 +339,16 @@ more than 2800 commits from 195 contributors since release 2.3.
 * We no longer release source tarballs. If you want to build from
   source, please build from the tags in the Git repository.
 
+* The Nix evaluator refuses to read from unregistered store paths,
+  to prevent reading garbage files.
+  This breaks builds that run the Nix evaluator in the sandbox using
+  an initially empty `NIX_STATE_DIR` but still using the regular store.
+  Builds that run the evaluator must now use a fresh store dir, where
+  sources can be written, registered and imported as expected.
+  See [#5495](https://github.com/NixOS/nix/issues/5495#issuecomment-964443016)
+  and [Nixpkgs#144197](https://github.com/NixOS/nixpkgs/commit/ca5ecbed4b5775f873fa15d7bf064b00d9e2862a)
+  for an example of a fix.
```

Comment:
I'm not sure that those fixes are great advise. People really shouldn't use `NIX_STATE_DIR` and `NIX_STORE_DIR`. It's better to use `--store dummy://` if you don't need a writable store (which should be the case for Nixpkgs), or `--store /some/path` if you do.

Reply:
I don't think `dummy` is sufficient, because it seems that the affected build needs to import from source; add a source (`builtins.path` etc) to the store and read back from it during evaluation.

Reply:
I think that points at a major bug in Nixpkgs. It should be possible to evaluate it in read-only mode. If that's not the case it has to be fixed.

Reply:
`--store /some/path` is not sufficient and `--readonly-mode` does not seem to affect the outcome either. See https://github.com/NixOS/nixpkgs/pull/145501

Reply:
Strange, I tried `nix-env -qa --drv-path --store dummy://` and it works.

Reply:
Well, it's still the same error.

Reply:
@edolstra the problem isn’t really about having a read-only store or not.

The issue here (in the linked issue I mean) is that the evaluation happens inside a derivation (doing some kind of read-only recursive Nix), and the files being evaluated are the inputs of the derivation (so are under `/nix/store`). And when the evaluator tries to access a path that’s under `storeDir`, it will assume that it’s meant as a store path, and so check its validity. Which fails inside the sandbox because although it has a view of a subset of the store, it doesn’t have access to the host’s Nix db to know that the path is indeed valid. Hence the error.

The reason for using `NIX_STORE_DIR` is just to prevent Nix from trying to check these paths (which look like store paths, but aren’t really). Using `--store 'dummy://?store=/somewhere/else'` would probably have the same effect